Added Scott's Allies03 co-op multiplayer mission as "Exodus"

Mailaender opened the pull request:
@ScottNZ's original mission converted from the [C# mission code found in release-20131223](https://github.com/OpenRA/OpenRA/blob/release-20131223/OpenRA.Mods.RA/Missions/Allies03Script.cs). This port is very faithful to the original. However I changed a lot of checks that were done every n ticks to use triggers instead, which didn't exist back then. As a result this should run pretty smooth.

Mailaender committed:
Port the Co-Op Allies03 map to Lua.

Mailaender committed:
Add a PassengerCount property to the Lua API.

Phrohdoh commented on OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs:
```diff
@@ -31,6 +31,9 @@ public TransportProperties(ScriptContext context, Actor self)
 		public bool HasPassengers { get { return cargo.Passengers.Any(); } }
 
+		[Desc("Specifies the amount of passengers.")]
+		public int Passengers { get { return cargo.Passengers.Count(); } }
```
This should be `PassengerCount`, just `Passengers` would make me think this is a table of actors.

Mailaender replied:
I somehow realized that as well when writing the commit message. Fixed.

r34ch commented:
Again can't comment on the code but the mission on hard is very easy. Perhaps the friendly reinforcements should be at intervals of units evacuated? The AI air units mostly circled their bases and AI attacks in general were mostly comprised of single units.

Mailaender commented:
I fixed the yak air raids. There were several problems I solved:
* `Attack` doesn't work without visibility so I added back the `CAMERA.Large`.
* `FlyCircle` is an activity that doesn't count as idle so the re-targeting never triggered.
* The yaks picked invalid targets. I restrained them to buildings and units now.

The constant yak harassment should increase difficulty.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+		end
+	end)
+
+	Trigger.OnAllKilled(SovietAirfields, function()
```
This wants to be `OnAllKilledOrCaptured`, I think.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+	Utils.Do(humans, function(player)
+		if player and player.IsLocalPlayer then
+			Trigger.AfterDelay(DateTime.Seconds(3), function()
+				Media.PlaySpeechNotification(allies1, "AlliedReinforcementsArrived")
```
`allies1` -> `player` I guess?

abcdefg30 replied:
I also found this a bit confusing, as I expected more reinforcements came in from another direction and I just didn't see them. Not sure how to fix that though, maybe reduce the delay to 2 seconds?

Mailaender replied:
Changed it to "Allied reinforcements have arrived from the North." to avoid confusion.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+{
+	hard = 200,
+	normal = 100,
+	easy = 50
```
This seems to increase the delay with difficulty. Intentional or swapped?

Mailaender replied:
Copy paste mistake: https://github.com/OpenRA/OpenRA/blob/release-20131223/OpenRA.Mods.RA/Missions/Allies03Script.cs#L408 has the values I wanted.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+
+	SovietParadrop = SovietParadrop + 1
+
+	Media.PlaySpeechNotification(allies1, "SovietForcesApproaching")
```
We'll want this for `allies2` as well, when he/she/it exists.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+
+AircraftTargets = function()
+	local targets = Utils.Where(Map.ActorsInWorld, function(a)
+		return a.Owner == allies1 and
```
`(a.Owner == allies1 or a.Owner == allies2) and ...`?

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+AircraftTargets = function()
+	local targets = Utils.Where(Map.ActorsInWorld, function(a)
+		return a.Owner == allies1 and
+			(a.HasProperty("StartBuildingRepairs") or a.HasProperty("Move"))
```
`a.HasProperty("Health")` would make more sense, imho. (What were those "invalid targets" you were talking about anyway?)

abcdefg30 replied:
Ah, I see. Yaks can't attack camo pillboxes. Those aren't excluded though, so maybe we want to check if the unit is cloaked?

Mailaender replied:
I just removed `HBOX` as that is easier and we are in a low tech mode anyway.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+	end)
+
+	unitsEvacuatedThreshold = UnitsEvacuatedThreshold[Difficulty]
+	UserInterface.SetMissionText(UnitsEvacuated .. "/" .. unitsEvacuatedThreshold .. " units evacuated.")
```
I think it would look better if we use the player color from the allied campaign here. Just visual and just me though.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,394 @@
+
+UnitsEvacuated = 0
+EvacuateAlliedUnit = function(unit)
+	if (unit.Owner == allies1 or unit.Owner == allies2) and unit.IsMobile then
```
This also wants a `unit..HasProperty("Move")` check.

Mailaender commented:
Thanks for the extensive review. Updated accordingly. Also removed a
```yaml
Buildable:
		Owner: allies, soviets
```
leftover from when we didn't have tech capturing. https://github.com/OpenRA/OpenRA/pull/8004

obrakmann commented:
I got this exception some time into the game:
```
Actor 'camera.paradrop' does not define a property 'Hunt'
```

obrakmann commented on OpenRA.Mods.Common/Scripting/Properties/AircraftProperties.cs:
```diff
@@ -60,5 +60,12 @@ public void Resupply()
+
+		[ScriptActorPropertyActivity]
+		[Desc("Returns true if the aircraft circling over it's home base.")]
+		public bool IsOnHoldingPattern
```
Just call it `IsCircling`. Also the description is misleading (the "over its home base" part).

obrakmann commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,398 @@
+	Utils.Do(SovietRallyPoints, function(rallyPoint)
+		local radius = WDist.FromCells(8)
+		local units = Map.ActorsInCircle(rallyPoint.CenterPosition, radius, function(actor)
+			return actor.Owner == soviets
```
I suppose adding `and actor.HasProperty("Hunt")` here would fix the crash I saw.

WolfGaming replied:
yeah this didn't crash before since every actor had the hunt property.

obrakmann commented on mods/ra/maps/exodus/rules.yaml:
```diff
@@ -0,0 +1,185 @@
+Player:
+	MissionObjectives:
+	PlayerResources:
+		DefaultCash: 5000
```
The whole `Player` block is redundant, it already has `MissionObjectives` and 5000 already is the `DefaultCash` default value.

obrakmann replied:
Err, slight correction. Because of [https://github.com/OpenRA/OpenRA/pull/11679/files#diff-c170037b524787407d1f07a0151d1d71R312](this check), you actually do want to set the `Cooperative` flag on `MissionObjectives`.

Background: You can leave the flag unset if you're sure that you will always change both players objectives the same way at the same time, but in the linked case, only one player's objective would be changed.

obrakmann replied:
Actually, the default cash thing was wrong, with the campaign rules, the default is 0. I'm afraid it needs to be there after all.

obrakmann commented:
Looks otherwise good to me. It's nice to have these back!

Mailaender commented:
Updated accordingly. Yeah, I got a bit nostalgic when looking at http://www.moddb.com/games/openra/images/evacuation#imagebox

obrakmann commented on mods/ra/maps/exodus/map.yaml:
```diff
@@ -0,0 +1,1429 @@
+		Owner: Soviets
+	Actor180: v01
+		Location: 100,29
+		Owner: Soviets
```
Is there a reason this church is owned by the Soviets?

Mailaender replied:
I assume it is a mistake in the original map design, a rather ironic one as well. Fixed.

obrakmann commented on OpenRA.Mods.Common/Scripting/Properties/AircraftProperties.cs:
```diff
@@ -60,5 +60,12 @@ public void Resupply()
 		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Returns true if the aircraft is flying circles.")]
```
Can we add that this is a hacky workaround for our sucky activities and that people shouldn't rely on it existing in the long term, please?

Mailaender replied:
Sounds a bit harsh, but sure.

obrakmann commented on mods/ra/maps/exodus/rules.yaml:
```diff
@@ -0,0 +1,180 @@
+			normal: Normal
+			hard: Hard
+		Default: normal
+
```
The debug menu can still be enabled.

Mailaender replied:
Why should we disallow it?

obrakmann replied:
I guess to prevent people from cheating on missions by being able to see the whole map. But of course they can now also just look at them in editor. I guess there's really no point to it then.

Could I ask you to add a commit that removes [these two lines](https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/maps/evacuation/rules.yaml#L6-L7) then, please?

Mailaender replied:
Done.

obrakmann commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,398 @@
+			return yak.CenterPosition.Z < 1 and yak.AmmoCount() == yak.MaximumAmmoCount()
+		end)
+		if #grounded > 0 then
+			Utils.Do(grounded, YakAttack)
```
This doesn't work because of [some weird code in `ResupplyAircraft`](https://github.com/OpenRA/OpenRA/blob/8233857be1e4684d0b7519760148404fecc0c8e8/OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs#L39). My suggestion would be to simply insert a `yak.Stop()` before `yak.Attack(target)` in `YakAttack`.

Mailaender replied:
Added the `yak.Stop()`.

obrakmann commented on OpenRA.Mods.Common/Scripting/Properties/AircraftProperties.cs:
```diff
@@ -8,7 +8,7 @@
  */
 #endregion
-
+ 
```
OpenRA.Mods.Common/Scripting/Properties/AircraftProperties.cs:L11: [SP2000] Invalid spacing at the end of the line.

obrakmann commented:
:+1:

Mailaender committed:
Don't force disable the debug menu for co-op missions.

Mailaender commented:
Fixed.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,399 @@
+AircraftTargets = function()
+	local targets = Utils.Where(Map.ActorsInWorld, function(a)
+		return (a.Owner == allies1 or a.Owner == allies2)
+			and (a.HasProperty("Move") or a.HasProperty("StartBuildingRepairs"))
```
Is there a reason why you don't just use `a.HasProperty("Health")` instead?

Mailaender replied:
I want them to snipe units first, then take on the buildings.

abcdefg30 replied:
I'm really unsure if this works that way.

obrakmann replied:
It doesn't. You'd have to sort the list returned by `Utils.Where` first.

abcdefg30 commented on mods/ra/maps/exodus/exodus.lua:
```diff
@@ -0,0 +1,399 @@
+		yak.Attack(target)
+	else
+		yak.ReturnToBase()
+		yak.Resupply()
```
This shouldn't be needed as pointed out by @obrakmann in https://github.com/OpenRA/OpenRA/pull/11658#discussion_r73580912.

Mailaender commented:
Fixed the naive code.

abcdefg30 commented:
This has my 👍 when https://github.com/Mailaender/OpenRA/pull/58 is accepted.

One final note though: Currently you are able to build in the evacuation area and thus are able to produce units *behind* the cell trigger. We could either expand the cell trigger or somehow try to prevent people from building there. But that's not a real blocker.

Mailaender commented:
Thanks for the code cleanup. I squashed them in. Removed https://github.com/Mailaender/OpenRA/pull/58/commits/c10b8c31a249200dcab0285b4670744aaed1f166 as that is decided by the script and I ditched https://github.com/Mailaender/OpenRA/pull/58/commits/74423648b8aedc16d82cbd235ea5ede43c968a60 and https://github.com/Mailaender/OpenRA/pull/58/commits/7af519b81e4066ee51a0437146c2ce7c8b312854 as that caused breakages in my tests.

abcdefg30 commented:
Bleh. I guess I won't block this PR (and release blocker) over this and file a follow-up PR.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ac1c9c61e8fc470bddbb95b89d7868c097fe5982...463f1f2cd6f00b2963b22f2865c95f04936bc6c7)
